fpga_interchange/site_router.cc

```diff
@@ -0,0 +1,753 @@
+        //
+        // The simplistic solution (only select when 1 solution is available)
+        // will likely solve initial problems.  Once that is show to be wrong,
+        // come back with something more general.
+        NPNR_ASSERT(false);
+
+    } while (!wire_to_expansion.empty());
+
+    return true;
+}
+
+bool Arch::SiteRouter::checkSiteRouting(const Context *ctx, const Arch::TileStatus &tile_status) const
+{
+    if (!dirty) {
+        return site_ok;
+    }
+
+    dirty = false;
+
+    if (cells_in_site.size() == 0) {
+        site_ok = true;
+        return site_ok;
+    }
+
+    site_ok = false;
+
+    // Make sure all cells in this site belong!
+    auto iter = cells_in_site.begin();
+    NPNR_ASSERT((*iter)->bel != BelId());
+    auto tile = (*iter)->bel.tile;
+
+    if (verbose_site_router(ctx)) {
+        log_info("Checking site routing for site %s\n",
+                 ctx->chip_info->sites[ctx->chip_info->tiles[tile].sites[site]].name.get());
+    }
+
+    for (CellInfo *cell : cells_in_site) {
+        // All cells in the site must be placed.
+        NPNR_ASSERT(cell->bel != BelId());
+
+        // Sanity check that all cells in this site are part of the same site.
+        NPNR_ASSERT(tile == cell->bel.tile);
+        NPNR_ASSERT(site == bel_info(ctx->chip_info, cell->bel).site);
+
+        // As a first pass make sure each assigned cell in site is valid by
+        // constraints.
+        if (!ctx->is_cell_valid_constraints(cell, tile_status, verbose_site_router(ctx))) {
+            if (verbose_site_router(ctx)) {
+                log_info("Sanity check failed, cell_type %s at %s has an invalid constraints, so site is not good\n",
+                         cell->type.c_str(ctx), ctx->nameOfBel(cell->bel));
+            }
+            site_ok = false;
+            return site_ok;
+        }
+    }
+    //
+    // FIXME: Populate "consumed_wires" with all VCC/GND tied in the site.
+    // This will allow route_site to leverage site local constant sources.
+    //
+    // FIXME: Handle case where a constant is requested, but use of an
+    // inverter is possible. This is the place to handle "bestConstant"
+    // (e.g. route VCC's over GND's, etc).
+    //
+    // FIXME: Enable some LUT rotation!
+    // Default cell/bel pin map always uses high pins, which will generate
+    // conflicts where there are none!!!
+
+    SiteInformation site_info(ctx, cells_in_site);
+
+    // Push from cell pins to the first WireId from each cell pin.
+    if (!site_info.check_initial_wires()) {
+        site_ok = false;
+        return site_ok;
+    }
+
+    site_ok = route_site(ctx, &site_info);
+    if (verbose_site_router(ctx)) {
+        if (site_ok) {
+            site_info.remove_routed_sources();
+            NPNR_ASSERT(site_info.is_fully_routed());
+            log_info("Site %s is routable\n",
+                     ctx->chip_info->sites[ctx->chip_info->tiles[tile].sites[site]].name.get());
```

Comment:
I think a `get_site_name(tile, site)` helper function might be a good idea here

Reply:
Fixed!